### PR TITLE
Add foreign compilation modes

### PR DESCRIPTION
**Pull request by snowleopard**
Add ability to make compilation of foreign stubs conditional on the mode via `(mode byte)` or `(mode native)`, which is useful for building the OCaml compiler with Dune.

For example:

```
(library
 (name foo)
 (modules)
 (foreign_stubs (language c) (mode byte) (flags -DBYTE) (names foo_bytes))
 (foreign_stubs (language c) (mode native) (flags -DNATIVE) (names foo_native)))

(executable
 (name main_byte)
 (modes byte)
 (libraries foo)
 (modules main_byte))

(executable
 (name main_native)
 (modes native)
 (libraries foo)
 (modules main_native))
```

where both `foo_bytes.c` and `foo_native.c` are defined as follows:

```c
#include <caml/mlvalues.h>
#ifdef BYTE
value foo_byte(value unit) { return Val_int(1); }
#endif
#ifdef NATIVE
value foo_native(value unit) { return Val_int(2); }
#endif
value foo_common(value unit) { return Val_int(3); }
```

Here is what we see in the test (instead of a failure due to multiple definitions of `foo_common`):

```
  $ cat >main_byte.ml <<EOF
  > external foo_byte : unit -> int = "foo_byte"
  > external foo_common : unit -> int = "foo_common"
  > let () = Printf.printf "%d" (foo_byte () + foo_common ())
  > EOF

  $ cat >main_native.ml <<EOF
  > external foo_native : unit -> int = "foo_native"
  > external foo_common : unit -> int = "foo_common"
  > let () = Printf.printf "%d" (foo_native () + foo_common ())
  > EOF

  $ ./sdune clean
  $ ./sdune build _build/default/main_byte.bc
  $ (cd _build/default; ocamlrun -I . main_byte.bc)
  4
  $ ./sdune exec ./main_native.exe
  5
```

/cc @mshinwell @jeremiedimino 

P.S.: There is also `Byte_with_stubs_statically_linked_in` link mode, which I'm not handling correctly.

### CODE REVIEW

**Comment by snowleopard** [comment]
@rgrinberg It would be interesting to know what you think about how this feature should interact with virtual libraries. 

Right now I simply include their objects in both static and dynamic archives, but it seems better to propagate the information about the compilation modes of their foreign objects somehow.

**Comment by nojb** [comment]
An alternative approach is for dune to simply export the mode as a define eg `DUNE_MODE_BYTE` or `DUNE_MODE_NATIVE` and have the stubs react to the define using `#ifdef`, etc.

**Comment by snowleopard** [comment]
@nojb Interesting idea, thanks! One problem with it is that it requires all stubs to be compiled twice, whereas the current draft changes compilation only of (hopefully) a small number of specially-defined stubs.

**Comment by rgrinberg** [comment]
>  It would be interesting to know how you think this feature should interact with virtual libraries.

It should work just fine.

> Right now I simply include all objects from virtual libraries into both static and dynamic archives, but it seems better to propagate the information about the compilation modes of their foreign objects somehow.

Yes, we'll need to propagate this information to the dune-package files.

> One problem with it is that it requires all stubs to be compiled twice, whereas the current draft changes compilation only of (hopefully) a small number of specially-defined stubs.

Is it possible to share sources between two foreign_stubs fields? If so, that would make it possible to use nojb's suggestion.


In the meeting we also discussed conditional OCaml sources for byte/native. Is this a requirement?

Also, it was discussed if was necessary to allow this for static vs. dynamic archives. Is this a requirement as well?

File: src/dune_rules/foreign.mli [hunk 1]
```diff
@@ -159,12 +172,35 @@ module Source : sig
...
+  (** Path to the object file in the build directory. *)
+  val build_path : t -> Path.t
+
+  (** Like [build_path] but returns [Some] only if the object should be included
+      into bytecode libraries. *)
+  val build_path_byte : t -> Path.t option
+
+  (** Like [build_path] but returns [Some] only if the object should be included
+      into native libraries. *)
+  val build_path_native : t -> Path.t option
```
Comment (rgrinberg) [inline src/dune_rules/foreign.mli:196]: We've become quite spartan with our public api's, so I'd suggest to trim helper functions that contain minimal logic. For example, I'd get rid of them all in favour of:

```
val obj_file : t -> Mode.t -> Path.Build.t option
```

The rest can be trivially accessed from the record fields.

**Comment by nojb** [comment]
Another possible approach is to enrich the `names` field of `foreign_stubs` as follows
```
(names (all foo_common) (byte foo_byte) (native foo_native))
```
where `(byte xxx)` means that `xxx` included for `byte` stubs and `(native xxx)` only for native, `(all xxx)` for all modes, etc. A plain name `xxx` is a shorthand for `(all xxx)`.

**Comment by hongchangwu** [comment]
I tried this with

```clojure
(executable
 (name main)
 (modes byte native)
 (foreign_stubs
  (language c)
  (mode native)
  (names stub)))
```

Still got

```shell
Error: Pure bytecode executables cannot contain foreign stubs.
Hint: If you only need to build a native executable use "(modes exe)".
```

I expect this error to not appear if I am only adding stubs in native mode.

**Comment by snowleopard** [comment]
@hongchangwu This was just an unfinished experiment, so I'm not surprised it doesn't fully work. I'm closing this draft PR to avoid confusion. Perhaps @mshinwell or @lukemaurer could say a few words since I believe they are using something similar to this commit in their fork https://github.com/ocaml-flambda/dune.

**Pull request closed without merging**